src/coreclr/jit/assertionprop.cpp

```diff
@@ -3104,23 +3104,51 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* tree)
         }
         break;
 
-#if FEATURE_HW_INTRINSICS
+#if FEATURE_SIMD
         case TYP_SIMD8:
+        {
+            if (!tree->OperIs(GT_LCL_VAR))
```

Comment:
Why the `!tree->OperIs(GT_LCL_VAR)` check?

Reply:
Ah, I forgot to add the comment.

If this is the child of something like a `GT_CALL` then we lose the type handle information and things complain. It's a much more involved change to either track the handle or to update the handling to look up the type from the method signature to know how its supposed to be passed.

The latter is probably the better "long term" change, IMO. As far as `TYP_SIMD16` is really concerned, its an opaque set of 128-bits. How that's interpreted depends entirely on the consuming operation.

We likewise have cases where `TYP_SIMD16` can represent a `Vector<T>`, a `Vector128<T>`, or a `Vector4` and where they are often "bitcast" between eachother. The source of truth for how that's passed around is dependent on the method signature, not on any type handle the JIT happens to be able to resolve or carry around.

Reply:
Presumably this problem can occur for any type of node then, not just `LCL_VAR`, i. e. this suggests it is not actually safe to constant propagate vector constants at all?

Though with that said, I wonder if you could "fix" this by just returning a faked-up handle from `gtGetStructHandle`. After morph we don't require precise handles (after all, we could propagate a `Vector4` local into a `Vector128<int>` one before).

Reply:
Possibly. There aren't many things that require some struct handle to be resolved. `GT_CALL` is one of the few.

I'm open to suggestions here. If we could detect the parent node we could limit propagation to specific contexts, but I couldn't determine how to do that here. Its really only important for a few hwintrinsic nodes that need to do containment here.

Reply:
> If we could detect the parent node

We could with `gtGetParent`, but it would be a (costly) hack.

The question here is how did the previous version of the code avoid the handle problem (I suspect by returning random handles) - we should match what it did.

I will pull the change down and see locally what sort of issues do we run into here.

Reply:
We tracked the relevant "base type" and looked up the handle from a cache. There were only a handful of instances where that can't actually resolve a handle.

Reply:
Yes, that is how it worked. The question is did we require the handle to actually be "correct" (related to the signature type) or not. My hypothesis is we did not and so we can return a "random" handle of matching size for `CNS_VEC`.